improved security, added user autocreation and manual user data sync from admin

MikhailRoot opened the pull request:
+added option to manually update user data on vanilla in users bulk
actions in wp-admin
+ added option to vanilla sso settings to automatically create user on
vanillaForum as it registers on wordpress
+ added needed direct php files access checks where needed and index.php
disallows to browse plugin directories.

DaazKu commented:
Tested.

DaazKu commented on admin.php:
```diff
@@ -114,7 +118,7 @@ function vf_admin_page() {
 		  url: '<?php echo site_url('wp-admin/admin.php?page=vf-admin-handle&vanillavalidate='); ?>'+validateUrl,
 		  success: function(data) {
-			 if (data.indexOf('http://') == 0 || data.indexOf('https://') == 0) {
+			 if (data.indexOf('http://') == 0 ||data.indexOf('https://') == 0) {
```
Put back the space.

DaazKu commented on admin.php:
```diff
@@ -171,9 +175,9 @@ function vf_sso_admin_page() {
 	$sso_clientid = vf_get_value('sso-clientid', $options, vf_format_url(get_option('blogname')));
-	$sso_secret = vf_get_value('sso-secret', $options, '');
-	$vanilla_url = vf_get_value('url', $options);
+	$sso_secret = vf_get_value('sso-secret', $options, '');$vanilla_url = vf_get_value('url', $options);
```
Put back the line feed.

DaazKu commented on admin.php:
```diff
@@ -249,9 +253,14 @@ function vf_sso_admin_page() {
 		<label>
 			<strong>Enable</strong>
-			<input type="checkbox" name="<?php echo vf_get_option_name('sso-enabled'); ?>" value="1"<?php echo $sso_enabled == '1' ? ' checked="checked"' : ''; ?> />
+			<input type="checkbox" name="<?php echo vf_get_option_name('sso-enabled'); ?>" value="1" <?php echo $sso_enabled == '1' ? ' checked="checked"' : ''; ?> />
```
You will end up with an extra space here.

DaazKu commented on admin.php:
```diff
@@ -280,20 +289,25 @@ function vf_sso_admin_page() {
+//			<!--</span></td>-->
+
+
+}
```
Better just to remove this completely.
Missing new line at EOF too.

DaazKu commented on force-user-sync.php:
```diff
@@ -0,0 +1,62 @@
+        'sso'=>$sso_string
+    );
+
+    $user_url=$base_vanilla_url.'entry/jsconnect?'.http_build_query($data);
```
You are missing a / between `$base_vanilla_url` and `'entry/jsconnect?'`

DaazKu commented on force-user-sync.php:
```diff
@@ -0,0 +1,62 @@
+    $options = get_option(VF_OPTIONS_NAME);
+    $base_vanilla_url = vf_get_value('url', $options, '');
+    $clientID = vf_get_value('sso-clientid', $options, '');
+    $sso_string=vf_get_sso_string(intval($user_id));
```
Missing spaces around the `=` sign

DaazKu commented on functions.php:
```diff
@@ -1,4 +1,8 @@
+//if ( ! defined( 'ABSPATH' ) ) {
+//	exit; // Exit if accessed directly.
+//}
+
```
Please do not add commented code.

DaazKu commented on functions.php:
```diff
@@ -165,9 +171,11 @@ function vf_validate_options($options) {
          $options['embed-matchcategories'] = $matchCategories;
 			break;
-		default:
+		default:{
```
This change should be undone.

DaazKu commented on functions.php:
```diff
@@ -177,22 +185,23 @@ function vf_validate_options($options) {
  * Validate that the provided url is a vanilla forum root. Returns properly formatted url if it is, or FALSE.
  */
 function vf_validate_url($url) {
```
You de-optimized that function.
The function was returning a result as soon as it reached `vf_parse_webroot()`.
Now it will do further check for nothing.
A `break` could solve that but that function was fine.

DaazKu commented:
The branch has conflict and the `sso-create-users-on-register` option does not work.
Please try to respect indenting and spacing.

MikhailRoot commented:
@DaazKu Thank you for help in making it better. i've updated it.

DaazKu commented on functions.php:
```diff
@@ -177,10 +180,11 @@ function vf_validate_options($options) {
  * Validate that the provided url is a vanilla forum root. Returns properly formatted url if it is, or FALSE.
  */
 function vf_validate_url($url) {
+	$result=false;
```
This line should be removed.

DaazKu commented on force-user-sync.php:
```diff
@@ -0,0 +1,62 @@
+        'sso'=>$sso_string
+    );
+
+    $user_url=$base_vanilla_url.'entry/jsconnect/?'.http_build_query($data);
```
Should be: 
`$user_url=$base_vanilla_url.'/entry/jsconnect?'.http_build_query($data);`

DaazKu commented:
Thanks you for taking the time to update your PR.
I added 2 more notes.

Could you also resolve the conflicts in your PR?
- See https://help.github.com/articles/syncing-a-fork/

MikhailRoot commented:
updated code in getting profile url, because it it's successfully got via parsing img tag, we don't need to get it again , otherwise we get it in alternate way. 
also updated user object checking because $cur_user->id!=='' will cause error if it's not WP_User - and this is what we really check for.

DaazKu commented:
The point of SSO is to NOT force-sync your users so we will not merge that PR.
